keras_rs/src/utils/pairwise_loss_utils.py

```diff
@@ -0,0 +1,100 @@
+from typing import Optional
+
+from keras import ops
+
+from keras_rs.src import types
+from keras_rs.src.utils.keras_utils import check_shapes_compatible
+
+
+def apply_pairwise_op(x: types.Tensor, op: ops) -> types.Tensor:
+    return op(
+        ops.expand_dims(x, axis=-1),
+        ops.expand_dims(x, axis=-2),
+    )
+
+
+def pairwise_comparison(
+    labels: types.Tensor,
+    logits: types.Tensor,
+    mask: types.Tensor,
+    logits_op: ops,
+) -> tuple[types.Tensor, types.Tensor]:
+    # Compute the difference for all pairs in a list. The output is a tensor
+    # with shape `(batch_size, list_size, list_size)`, where `[:, i, j]` stores
+    # information for pair `(i, j)`.
+    pairwise_labels_diff = apply_pairwise_op(labels, ops.subtract)
+    pairwise_logits = apply_pairwise_op(logits, logits_op)
+
+    # Keep only those cases where `l_i < l_j`.
+    pairwise_labels = ops.cast(
+        ops.greater(pairwise_labels_diff, 0), dtype=labels.dtype
+    )
+    if mask is not None:
+        valid_pairs = apply_pairwise_op(mask, ops.logical_and)
+        pairwise_labels = ops.multiply(
+            pairwise_labels, ops.cast(valid_pairs, dtype=pairwise_labels.dtype)
+        )
+
+    return pairwise_labels, pairwise_logits
+
+
+def process_loss_call_inputs(
+    y_true: types.Tensor,
+    y_pred: types.Tensor,
+    mask: Optional[types.Tensor] = None,
+) -> tuple[types.Tensor, types.Tensor, Optional[types.Tensor]]:
+    """
+    This utility function does three things:
+
+    - Checks that `y_true`, `y_pred` are of rank 1 or 2.
+    - Check that `y_true`, `y_pred`, `mask` have the same shape.
+    - Add batch dimension if rank = 1.
+    """
+
+    def get_shape(x: types.Tensor) -> tuple[types.TensorShape, int]:
```

Comment:
I don't think this is worth the trouble (and the name is a bit confusing).

You can just do `len(y_true_shape)` where needed.